File: Sources/AudioLocal/ImportedBook.swift
import Foundation

struct ImportedBook: Identifiable {
    let id = UUID()
    var title: String
    var author: String?
    let sourceURL: URL
    var chapters: [ImportedChapter]

    var totalWordCount: Int {
        chapters.reduce(0) { $0 + $1.wordCount }
    }

    var includedChapters: [ImportedChapter] {
        chapters.filter(\.isIncluded)
    }

    var includedWordCount: Int {
        includedChapters.reduce(0) { $0 + $1.wordCount }
    }
}

struct ImportedChapter: Identifiable, Hashable {
    let id: UUID
    let importOrder: Int
    var title: String
    let originalText: String
    var workingText: String
    var isIncluded: Bool
    let sourcePath: String

    init(
        id: UUID = UUID(),
        importOrder: Int,
        title: String,
        originalText: String,
        workingText: String? = nil,
        isIncluded: Bool = true,
        sourcePath: String
    ) {
        self.id = id
        self.importOrder = importOrder
        self.title = title
        self.originalText = originalText
        self.workingText = workingText ?? originalText
        self.isIncluded = isIncluded
        self.sourcePath = sourcePath
    }

    var wordCount: Int {
        workingText.split { $0.isWhitespace || $0.isNewline }.count
    }

    var hasEdits: Bool {
        workingText != originalText
    }
}

File: Windows/src/AudioLocal.Windows.Core/Models/ImportedBook.cs
namespace AudioLocal.Windows.Core.Models;

public sealed class ImportedBook
{
    public required string Title { get; init; }
    public string? Author { get; init; }
    public required string SourcePath { get; init; }
    public required IReadOnlyList<ImportedChapter> Chapters { get; init; }
    public byte[]? CoverImageData { get; init; }
    public string? CoverMimeType { get; init; }

    public int TotalWordCount => Chapters.Sum(static chapter => chapter.WordCount);

    public IReadOnlyList<ImportedChapter> IncludedChapters =>
        Chapters.Where(static chapter => chapter.IsIncluded).ToArray();
}

public sealed class ImportedChapter
{
    public required Guid Id { get; init; }
    public required int ImportOrder { get; init; }
    public required string Title { get; set; }
    public required string OriginalText { get; init; }
    public required string WorkingText { get; set; }
    public required string SourcePath { get; init; }
    public bool IsIncluded { get; set; } = true;

    public int WordCount => string.IsNullOrWhiteSpace(WorkingText)
        ? 0
        : WorkingText.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries).Length;

    public bool HasEdits => !string.Equals(WorkingText, OriginalText, StringComparison.Ordinal);
}
